### PR TITLE
Imported avoiding lockin story.

### PR DESCRIPTION
**Pull request by mooreds**

### CODE REVIEW

File: site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+---
+layout: blog-post
+title: Avoid Lock-In by Insulating Your Application and Having a Backup Plan
```
Comment (dpolstra) [inline site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md:3]: How to protect your organization from auth vendor lock-in

File: site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+How can you avoid this? 
+
+Two strategies that work are:
+
+* Insulate Your Application
+* Have a Backup Plan
+
+_This blog post is an excerpt from [Avoiding Authentication System Lock-in](/learn/expert-advice/identity-basics/avoid-lockin/)._
+
+## Insulate your application
```
Comment (dpolstra) [inline site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md:25]: "Isolate your application"?

Comment (dpolstra) [inline site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md:25]: or "abstract your application"

File: site/learn/expert-advice/identity-basics/avoid-lockin.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+If you end up going with a provider who doesn't offer full-data exports, you may be in trouble down the line. Data like Rules and Roles might need to be manually reconfigured in most cases, or scripted with [something like Terraform](https://registry.terraform.io/providers/alexkappa/auth0/latest/docs). User data - including password hashes - will be much more problematic if you cannot get it out of the system when you need it, so be sure to check on this before making a decision.
+
+## Limit Your Usage Where You Can
+
+Regardless of which provider you go with, many of them offer the ability to store extra data in their system. Being able to store things like user properties and roles on their system seems nice at first, but might be more trouble than it's worth. Generally, it is better to keep details about your user and their relation to your application, in your application. You will doubtlessly have a user model in your system, and this is a better place to store these details (or on related models) because it gives you an extra layer of insulation from the auth provider.
```
Comment (dpolstra) [inline site/learn/expert-advice/identity-basics/avoid-lockin.md:40]: extra layer of isolation

File: site/learn/expert-advice/identity-basics/avoid-lockin.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+## Limit Your Usage Where You Can
+
+Regardless of which provider you go with, many of them offer the ability to store extra data in their system. Being able to store things like user properties and roles on their system seems nice at first, but might be more trouble than it's worth. Generally, it is better to keep details about your user and their relation to your application, in your application. You will doubtlessly have a user model in your system, and this is a better place to store these details (or on related models) because it gives you an extra layer of insulation from the auth provider. 
+
+Even if you might want to filter authorization to an application so that only your employees can access it, this would be an excellent time to use provider-side attributes to flag authorized users. In this scenario, you'd likely have a better time adding authorized users to a group in your enterprise identity system (such as a business Google or Microsoft account) and asserting that property in your auth system. This way, even if you need to move to another provider later, your user attributes will remain intact without needing to be migrated.
+
+## Insulate Your Application
```
Comment (dpolstra) [inline site/learn/expert-advice/identity-basics/avoid-lockin.md:44]: see above

File: site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+When implementing an authentication system (and various other services provided by third-party vendors, such as storage), there is usually an integration involved. This integration requires writing code in your application to interact with the vendor's service.
+
+It is usually a good idea to write this code so that the vendor is insulated away from your application code. Often, vendors will provide an SDK for interfacing with their service. By wrapping the usage of this SDK and abstracting it with a more generic interface, we can have greater control over how the vendor ties into our application.
+
+If the vendor uses Open Standards as described above, this makes things easier. We can build an interface that describes how we would ideally like to interact with our authentication service and then implement this interface with a wrapper around the vendor's SDK. In the future, if things go south and we need to break away from the vendor, we can just re-implement this interface with an integration for a new provider, and with any luck, things will transition smoothly.
```
Comment (jobannon) [inline site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md:31]: In the future, if things go south and we need to break away from the vendor, we can just re-implement this interface with an integration for a new provider, and with any luck, things will transition smoothly.

This seems like a long sentence

File: site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+{% include _image.liquid src="/assets/img/advice/avoid-lockin/facade-diagram.jpg" alt="Diagram of Authentication Interface example." class="img-fluid" figure=false %}
+
+As you can see in this simple diagram, by having your core services rely on a generic abstraction of an authentication system rather than a concrete implementation of one, you can retain the freedom to reimplement the interface with code to communicate with whatever provider you like. If you built a concrete dependency on one particular provider instead, you would have lots of refactoring and retesting to do when switching implementations. Naturally, even with a well-abstracted interface, you will still need thorough testing.
```
Comment (jobannon) [inline site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md:35]: As you can see in this simple diagram, by having your core services rely on a generic abstraction of an authentication system rather than a concrete implementation of one, you can retain the freedom to reimplement the interface with code to communicate with whatever provider you like.

This sentence is hard to parse/read.

File: site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+It's crucial to build integration tests around this kind of abstraction. Having integration tests in place to give you confidence in your abstraction and interface is always good, but especially so when there is the possibility of re-implementing the underlying code at some point. As long as the new implementation holds to the interface, you will have a good level of confidence that things are working as expected and that your application still works after a migration.
```
Comment (jobannon) [inline site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md:37]: `that your abstraction and interface is always good`

File: site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+## Have a backup plan
+
+Generally, you don't implement one solution while actively planning to swap it out in the future. Because of this, it is easy to accidentally put on the proverbial blinders and only focus on what is right in front of you without seeing the bigger picture and the long-term ramifications of certain design decisions. 
+
+This can lead to situations where implementations become too specific to the current vendor and are prohibitively expensive to change in the future. "If only we had known we were going to switch providers, we would have done things differently." Factoring in and budgeting for a Plan B or future migration can mitigate such tunnel vision
```
Comment (jobannon) [inline site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md:43]: missing period

File: site/learn/expert-advice/identity-basics/avoid-lockin.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+This is a similar principle to the test-first mindset. If you write your code with testing as a foremost concern, the output will likely be different - and more testable by nature - than if you conducted testing as an afterthought. Making architectural decisions while mindfully being aware of possible future migrations can lead to more robust system design and easier migrations in the future.
+
+
+## Wrapping Up
+
+There will always be the potential for issues when switching out something as central to your application as Authentication , no matter how much you plan and prepare in advance. This risk can be reduced or eliminated by following practices like those described above to protect your application from the uncertainty inherent to third-party integrations. The value that using a service like this brings is not to be understated, as long as you understand the risks and take appropriate measures to minimize them.
```
Comment (jobannon) [inline site/learn/expert-advice/identity-basics/avoid-lockin.md:73]: central to your application as Authentication <space>, ?

Comment (jobannon) [reply site/learn/expert-advice/identity-basics/avoid-lockin.md:73]: space

File: site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md (hunk 1 above)
Comment (mooreds) [reply site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md:3]: Love it!

**Review by jobannon — APPROVED** [review]
Interesting Read!

File: site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md (hunk 2 above)
Comment (mooreds) [reply site/_posts/2021-07-03-avoid-lockin-insulate-your-application-and-have-backup-plan-excerpt.md:25]: Insulate is a better fit. "insulate your application from the API of your auth provider" makes more sense than "isolate your application from the auth provider" to me.
